### PR TITLE
Add the support of the BF.MEXISTS and BF.CARD Command

### PR DESCRIPTION
**Pull request by zncleon**
This PR introduce the bloom BF.MEXISTS and BF.CARD command.

[1] [https://redis.io/commands/bf.mexists/](https://redis.io/commands/bf.mexists/)
[2] [https://redis.io/commands/bf.card/](https://redis.io/commands/bf.card/)

### CODE REVIEW

File: src/commands/cmd_bloom_filter.cc [hunk 1]
```diff
@@ -105,14 +105,37 @@ class CommandBFExists : public Commander {
...
 
     *output = redis::Integer(ret);
     return Status::OK();
   }
 };
 
+class CommandBFMExists : public Commander {
+ public:
+  Status Parse(const std::vector<std::string> &args) override {
+    for (size_t i = 2; i < args_.size(); ++i) {
```
Comment (mapleFU) [inline src/commands/cmd_bloom_filter.cc:120]: reserve before emplace

File: src/types/redis_bloom_chain.cc [hunk 2]
```diff
@@ -185,31 +185,40 @@ rocksdb::Status BloomChain::Add(const Slice &user_key, const Slice &item, int *r
...
-    s = bloomCheck(bf_key_list[i], item_string, &exist);
-    if (!s.ok()) return s;
-    if (exist) {
-      break;
+  for (size_t i = 0; i < items.size(); ++i) {
+    std::string item_string = items[i].ToString();
+    // check
+    bool exist = false;
+    for (int ii = metadata.n_filters - 1; ii >= 0; --ii) {  // TODO: to test which direction for searching is better
+      s = bloomCheck(bf_key_list[ii], item_string, &exist);
```
Comment (mapleFU) [inline src/types/redis_bloom_chain.cc]: Can we support `bloomCheck` with vector? Otherwise, bloom filter will be get multiple times, which might not good for LRU and BlockCache

File: src/types/redis_bloom_chain.cc [hunk 3]
```diff
@@ -105,9 +106,14 @@ rocksdb::Status BloomChain::bloomCheck(const Slice &bf_key, const std::string &i
...
 
-  uint64_t h = BlockSplitBloomFilter::Hash(item.data(), item.size());
-  *exist = block_split_bloom_filter.FindHash(h);
-
+  for (size_t i = 0; i < items.size(); ++i) {
+    if (exists->at(i)) {
+      continue;
+    }
+    std::string item = items[i].ToString();
+    uint64_t h = BlockSplitBloomFilter::Hash(item.data(), item.size());
```
Comment (mapleFU) [inline src/types/redis_bloom_chain.cc]: Why using `ToString` to copying here? Slice also has get `const char*` and length

File: src/types/redis_bloom_chain.cc [hunk 4]
```diff
@@ -105,9 +106,14 @@ rocksdb::Status BloomChain::bloomCheck(const Slice &bf_key, const std::string &i
   BlockSplitBloomFilter block_split_bloom_filter;
   block_split_bloom_filter.Init(std::move(bf_data));
 
-  uint64_t h = BlockSplitBloomFilter::Hash(item.data(), item.size());
-  *exist = block_split_bloom_filter.FindHash(h);
-
+  for (size_t i = 0; i < items.size(); ++i) {
+    if (exists->at(i)) {
```
Comment (mapleFU) [inline src/types/redis_bloom_chain.cc]: Why? this is pre-checked before calling `blockCheck`?

File: tests/gocase/unit/type/bloom/bloom_test.go [hunk 5]
```diff
@@ -253,6 +268,18 @@ func TestBloom(t *testing.T) {
...
+		require.NoError(t, rdb.Del(ctx, "no_exist_key").Err())
+		require.Equal(t, int64(0), rdb.Do(ctx, "bf.card", "no_exist_key").Val())
+
+		require.NoError(t, rdb.Del(ctx, key).Err())
+		require.NoError(t, rdb.Do(ctx, "bf.reserve", key, "0.02", "1000", "expansion", "1").Err())
+		require.Equal(t, int64(0), rdb.Do(ctx, "bf.card", key).Val())
+		require.NoError(t, rdb.Do(ctx, "bf.add", key, "item1").Err())
+		require.Equal(t, int64(1), rdb.Do(ctx, "bf.card", key).Val())
+		require.NoError(t, rdb.Do(ctx, "bf.add", key, "item2").Err())
+		require.Equal(t, int64(2), rdb.Do(ctx, "bf.card", key).Val())
```
Comment (mapleFU) [inline tests/gocase/unit/type/bloom/bloom_test.go:285]: Should we add a test for insert duplicate key?

**Review by mapleFU — COMMENTED** [review]
Rest looks ok to me

File: src/types/redis_bloom_chain.cc [hunk 6]
```diff
@@ -144,21 +150,19 @@ rocksdb::Status BloomChain::Add(const Slice &user_key, const Slice &item, int *r
...
   batch->PutLogData(log_data.Encode());
 
-  std::string item_string = item.ToString();
-
   // check
-  bool exist = false;
+  std::vector<bool> exist(1, false);                   // TODO: to refine in BF.MADD
   for (int i = metadata.n_filters - 1; i >= 0; --i) {  // TODO: to test which direction for searching is better
-    s = bloomCheck(bf_key_list[i], item_string, &exist);
+    s = bloomCheck(bf_key_list[i], {item}, &exist);
```
Comment (git-hulk) [inline src/types/redis_bloom_chain.cc:156]: It seems likely inconsistent return type between bloomCheck and MExists, the first one is vector<bool> and the later is vector<int>

File: src/types/redis_bloom_chain.h [hunk 7]
```diff
@@ -63,7 +64,11 @@ class BloomChain : public Database {
                                    BloomChainMetadata *metadata);
   void createBloomFilterInBatch(const Slice &ns_key, BloomChainMetadata *metadata,
                                 ObserverOrUniquePtr<rocksdb::WriteBatchBase> &batch, std::string *bf_data);
+
+  /// bf_data: [in/out] The content string of bloomfilter.
   static void bloomAdd(const std::string &item, std::string *bf_data);
-  rocksdb::Status bloomCheck(const Slice &bf_key, const std::string &item, bool *exist);
+
+  /// exists: [in/out] The items exist in bloomfilter already or not.
```
Comment (mapleFU) [inline src/types/redis_bloom_chain.h:71]:
```suggestion
  /// exists: [out] The items exist in bloomfilter already or not.
```

File: src/types/redis_bloom_chain.cc (hunk 4 above)
Comment (zncleon) [reply src/types/redis_bloom_chain.cc]: I think if the item exists in other sub-bloomfilter already, It's not necessary to check in this sub-bloomfilter.

File: src/types/redis_bloom_chain.cc (hunk 6 above)
Comment (zncleon) [reply src/types/redis_bloom_chain.cc:156]: Sure, I think I will support `MAdd` in next patch, the return type of `MAdd` will be vector. And the `Add` functions will call `MAdd`

File: src/types/redis_bloom_chain.cc (hunk 4 above)
Comment (mapleFU) [reply src/types/redis_bloom_chain.cc]: Oh I see, you can add a comment for this. Since it would be a bit confusing

File: src/types/redis_bloom_chain.cc [hunk 8]
```diff
@@ -105,9 +106,14 @@ rocksdb::Status BloomChain::bloomCheck(const Slice &bf_key, const std::string &i
   BlockSplitBloomFilter block_split_bloom_filter;
   block_split_bloom_filter.Init(std::move(bf_data));
 
-  uint64_t h = BlockSplitBloomFilter::Hash(item.data(), item.size());
-  *exist = block_split_bloom_filter.FindHash(h);
-
+  for (size_t i = 0; i < items.size(); ++i) {
+    // this item exists in other bloomfilter already, and it's not necessary to check in this bloomfilter.
+    if (exists->at(i)) {
```
Comment (PragmaTwice) [inline src/types/redis_bloom_chain.cc]: `at` will throw exceptions, so you'd better to use `operator[]` here.

File: src/types/redis_bloom_chain.cc (hunk 6 above)
Comment (git-hulk) [reply src/types/redis_bloom_chain.cc:156]: OK, makes sense.

File: src/types/redis_bloom_chain.h (hunk 7 above)
Comment (mapleFU) [reply src/types/redis_bloom_chain.h:71]: Oh I misunderstand this, so `exists` is a `[in/out]` arg. ignore this

File: src/server/redis_reply.h [hunk 9]
```diff
@@ -40,6 +40,15 @@ std::string Integer(T data) {
   return ":" + std::to_string(data) + CRLF;
 }
 
+template <typename T, std::enable_if_t<std::is_integral_v<T>, int> = 0>
+std::string MultiInteger(const std::vector<T> &multi_data) {
```
Comment (mapleFU) [inline src/server/redis_reply.h:44]: A quick question: would other Redis response can be help with `MultiInteger`?

File: src/types/redis_bloom_chain.cc [hunk 10]
```diff
@@ -105,9 +106,14 @@ rocksdb::Status BloomChain::bloomCheck(const Slice &bf_key, const std::string &i
...
-  uint64_t h = BlockSplitBloomFilter::Hash(item.data(), item.size());
-  *exist = block_split_bloom_filter.FindHash(h);
-
+  for (size_t i = 0; i < items.size(); ++i) {
+    // this item exists in other bloomfilter already, and it's not necessary to check in this bloomfilter.
+    if ((*exists)[i]) {
+      continue;
+    }
+    uint64_t h = BlockSplitBloomFilter::Hash(items[i].data(), items[i].size());
+    exists->at(i) = block_split_bloom_filter.FindHash(h);
```
Comment (mapleFU) [inline src/types/redis_bloom_chain.cc]:
```suggestion
    exists[i] = block_split_bloom_filter.FindHash(h);
```

File: src/types/redis_bloom_chain.cc [hunk 11]
```diff
@@ -185,31 +189,37 @@ rocksdb::Status BloomChain::Add(const Slice &user_key, const Slice &item, int *r
   return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
 }
 
-rocksdb::Status BloomChain::Exist(const Slice &user_key, const Slice &item, int *ret) {
+rocksdb::Status BloomChain::Exists(const Slice &user_key, const Slice &item, int *ret) {
+  std::vector<int> tmp(1, 0);
```
Comment (mapleFU) [inline src/types/redis_bloom_chain.cc]: Nit
```suggestion
  std::vector<int> tmp{0};
```

Would this be better to understand?

File: src/server/redis_reply.h (hunk 9 above)
Comment (mapleFU) [reply src/server/redis_reply.h:44]: (Oh and @PragmaTwice and can we support a `vector<bool>` version to convert `vector<bool>` to 0/1 response?)

**Review by mapleFU — DISMISSED** [review]
LGTM

File: src/types/redis_bloom_chain.cc [hunk 12]
```diff
@@ -185,31 +189,37 @@ rocksdb::Status BloomChain::Add(const Slice &user_key, const Slice &item, int *r
...
+    s = bloomCheck(bf_key_list[i], items, &exists);
     if (!s.ok()) return s;
-    if (exist) {
-      break;
-    }
   }
-  *ret = exist ? 1 : 0;
+
+  for (size_t i = 0; i < items.size(); ++i) {
+    rets->at(i) = exists[i] ? 1 : 0;
```
Comment (PragmaTwice) [inline src/types/redis_bloom_chain.cc]:
```suggestion
    (*rets)[i] = exists[i] ? 1 : 0;
```

File: src/server/redis_reply.h (hunk 9 above)
Comment (PragmaTwice) [reply src/server/redis_reply.h:44]: not sure since resp3 supports boolean.

Comment (zncleon) [reply src/server/redis_reply.h:44]: It's seems only here(`BF.MExists` and `BF.MAdd`) and `SMIsMember` can be help with `MultiInteger`. Should we give up `MultiInteger`?

Comment (mapleFU) [reply src/server/redis_reply.h:44]: You can put it in `src/commands/cmd_bloom_filter.cc` in anonymous namespace as a helper first. If you find any other user would use it, you can move it out.

Comment (PragmaTwice) [reply src/server/redis_reply.h:44]: this function looks ok to me.

**Review by PragmaTwice — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Review by mapleFU — APPROVED** [review]
